### PR TITLE
More couchform refactors

### PR DESCRIPTION
**Pull request by benrudolph**
@snopoke more data access abstraction. trying to separate out locking from data access. a few other minor tweaks as well. best read cbc

cc: @esoergel

### CODE REVIEW

File: corehq/ex-submodules/casexml/apps/case/tests/test_rebuild.py [hunk 1]
```diff
@@ -380,9 +380,9 @@ def testArchiveModifiedOn(self):
         way_earlier = now - timedelta(days=1)
         # make sure we timestamp everything so they have the right order
         create_block = CaseBlock(case_id, create=True, date_modified=way_earlier)
-        FormProcessorInterface.post_case_blocks([create_block.as_xml()], form_extras={'received_on': way_earlier})
+        FormProcessorInterface().post_case_blocks([create_block.as_xml()], form_extras={'received_on': way_earlier})
```
Comment (dimagimon) [inline corehq/ex-submodules/casexml/apps/case/tests/test_rebuild.py]: E501 line too long (116 > 115 characters)

File: corehq/ex-submodules/couchforms/__init__.py [hunk 2]
```diff
@@ -1,3 +1,3 @@
 from .exceptions import XMLSyntaxError, CouchFormException
 from .getters import *
-from .util import SubmissionPost, fetch_and_wrap_form, convert_xform_to_json
+from .util import SubmissionPost, fetch_and_wrap_form
```
Comment (dimagimon) [inline corehq/ex-submodules/couchforms/__init__.py]: F401 'SubmissionPost' imported but unused

Comment (dimagimon) [inline corehq/ex-submodules/couchforms/__init__.py]: F401 'fetch_and_wrap_form' imported but unused

**Comment by esoergel** [comment]
@benrudolph Can we sit down tomorrow and pair review a bit of this stuff?  I feel like I've got a sense of what's going on, but it'd be nice to get the full picture.

**Comment by benrudolph** [comment]
@esoergel yep no problem. be nice for me to bounce ideas off as well.

will fix tests though one of the failures looks unrelated

File: corehq/ex-submodules/casexml/apps/phone/tests/test_ota_restore.py [hunk 3]
```diff
@@ -122,7 +122,8 @@ def testUserRestoreWithCase(self):
             xml_data = f.read()
 
         # implicit length assertion
-        _, _, [newcase] = FormProcessorInterface.submit_form_locally(xml_data, domain=self.project.name)
+        _, _, [newcase] = FormProcessorInterface().submit_form_locally(xml_data, domain=self.project.name)
+        user = dummy_user()
```
Comment (dimagimon) [inline corehq/ex-submodules/casexml/apps/phone/tests/test_ota_restore.py]: F841 local variable 'user' is assigned to but never used

File: corehq/ex-submodules/couchforms/__init__.py (hunk 2 above)
Comment (dimagimon) [inline corehq/ex-submodules/couchforms/__init__.py]: F401 'SubmissionPost' imported but unused

Comment (dimagimon) [inline corehq/ex-submodules/couchforms/__init__.py]: F401 'fetch_and_wrap_form' imported but unused

**Comment by benrudolph** [comment]
@snopoke had a good discussion with @esoergel about this today. i'm thinking we can with a new kind of schema that will hopefully make it easier to do this. it's kind of like the polymorphism you were talking about today.

Models:

`AbstractGenericXForm` - this model provides the methods that an xform, either sql or couch, need to define.
`Generic<Couch|SQL>XForm` - this model inherits the `AbstractGenericXForm` and implements appropriate methods. It also has a point to the actual model, i.e. `XFormInstance` under something like `self.xform`. This way we can keep the generic versions clean and have only the methods needed to implement form processing. 

Interfaces:

`FormProcessorInterface` - this interface will still implement the larger operation on xforms and cases. will return either a `GenericCouchXForm` or `GenericSQLXForm`

Note: The other interfaces, `XFormInterface`, `CaseInterface`, and `SynclogInterface` i'm planning on getting rid of.

I'm hoping this will allow us a little more freedom to partially roll out the sql models. What do you think?

**Comment by snopoke** [comment]
:+1: That sounds good. I think that's basically the adapter pattern I was talking about with an abstract class that defines the common interface

**Pull request merged**